src/util/codec/table.rs

```diff
@@ -228,44 +228,118 @@ pub trait TableDecoder: DatumDecoder {
 }
 
 impl<T: BytesDecoder> TableDecoder for T {}
+pub type ValDict<'a> = HashMap<i64, &'a [u8]>;
 
-// `cut_row` cut encoded row into byte slices and return interested columns' byte slice.
-// Row layout: colID1, value1, colID2, value2, .....
-pub fn cut_row<'a>(mut data: &'a [u8], cols: &HashSet<i64>) -> Result<ValDict<'a>> {
-    // hack: HashMap will still allocate memeory when capacity is 0, need to use new instead.
+pub struct RowColMeta {
+    offset: usize,
+    length: usize,
+}
+
+pub struct RowColsDict {
+    value: Vec<u8>,
+    cols: HashMap<i64, RowColMeta>,
+}
+
+impl RowColMeta {
+    pub fn new(offset: usize, length: usize) -> RowColMeta {
+        RowColMeta {
+            offset: offset,
+            length: length,
+        }
+    }
+}
+
+impl RowColsDict {
+    pub fn new(cols_num: usize, val: Vec<u8>) -> RowColsDict {
+        RowColsDict {
+            value: val,
+            cols: {
+                if cols_num == 0 {
+                    HashMap::new()
+                } else {
+                    HashMap::with_capacity(cols_num)
+                }
+            },
+        }
+    }
+
+    pub fn insert(&mut self, key: i64, offset: usize, length: usize) -> Option<RowColMeta> {
+        let value = RowColMeta::new(offset, length);
+        self.cols.insert(key, value)
+    }
+
+    pub fn len(&self) -> usize {
+        self.cols.len()
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.cols.is_empty()
+    }
+
+    pub fn get(&self, key: &i64) -> Option<&[u8]> {
+        if let Some(meta) = self.cols.get(key) {
+            return Some(&self.value.as_slice()[meta.offset..(meta.offset + meta.length)]);
+        }
+        None
+    }
+
+    pub fn to_hash_map(&self) -> HashMap<i64, Vec<u8>> {
+        if self.cols.is_empty() {
+            return HashMap::new();
+        }
+        let mut data = HashMap::with_capacity(self.cols.len());
+        for key in self.cols.keys() {
```

Comment:
use `for (key, meta) in &self.cols` to get the key, meta and corresponding data directly, no need to call `self.get` to get again.